### PR TITLE
Feature/is nan

### PR DESCRIPTION
**Pull request by reyoung**
This is a part of #7092 .

### CODE REVIEW

File: paddle/framework/tensor_util.cc [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+  AnyDTypeVisitor(Predicate predicate, const Tensor& tensor, const DevCtx& ctx,
+                  Tensor* out)
+      : predicate_(predicate), tensor_(tensor), ctx_(ctx), out_(out) {}
+
+  template <typename T>
+  void operator()() const {
+    auto t = EigenVector<T>::Flatten(tensor_);
+    auto o = EigenScalar<bool>::From(*out_);
+    o.device(*ctx_.eigen_device()) = predicate_(t).any();
```
Comment (QiJune) [inline paddle/framework/tensor_util.cc]: We can add a reference comment on eigen any() for better understanding
https://eigen.tuxfamily.org/dox/classEigen_1_1DenseBase.html#abfbf4cb72dd577e62fbe035b1c53e695

File: paddle/framework/tensor_util.cc [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  AnyVisitor<Predicate> visitor(tensor, predicate);
+  auto place = tensor.place();
+  return platform::VisitPlace(place, visitor);
+}
+
+struct HasNANPredicate {
+  template <typename T>
+  auto operator()(const T& eigen_vec) const
+      -> decltype(std::declval<T>().isnan()) {
+    return eigen_vec.isnan();
```
Comment (QiJune) [inline paddle/framework/tensor_util.cc]: add a reference comment:
https://eigen.tuxfamily.org/dox/classEigen_1_1ArrayBase.html#aab10b156cb69206461728782dd01c37a

File: paddle/framework/tensor_util.cc [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+  }
+
+  bool GetResult(const framework::Tensor& out,
+                 const platform::CUDAPlace& gpu) const {
+    platform::CPUPlace cpu;
+    framework::Tensor tmp;
+    tmp.Resize({1});
+    tmp.mutable_data<bool>(cpu);
+    platform::DeviceContextPool::Instance().Get(gpu)->Wait();
+    CopyFrom(out, cpu, &tmp);
```
Comment (QiJune) [inline paddle/framework/tensor_util.cc]: If we do copy between CPU and GPU, we need to pass a DeviceContext

File: paddle/framework/tensor_util_test.cc [hunk 4]
```diff
@@ -230,5 +231,28 @@ TEST(CopyToVector, Tensor) {
 #endif
 }
 
+TEST(IsNAN, CPU) {
```
Comment (QiJune) [inline paddle/framework/tensor_util_test.cc]: Please add GPU unit tests at the same time.

File: paddle/platform/device_context.h [hunk 5]
```diff
@@ -52,6 +52,14 @@ class CPUDeviceContext : public DeviceContext {
   std::unique_ptr<Eigen::DefaultDevice> eigen_device_;
 };
 
+template <typename Place>
+struct DefaultDeviceContextType;
```
Comment (QiJune) [inline paddle/platform/device_context.h]: What is the `DefaultDeviceContextType ` used for?

File: paddle/framework/tensor_util.cc (hunk 3 above)
Comment (reyoung) [reply paddle/framework/tensor_util.cc]: However, the device context is a global variable. We can get DeviceContext by its place.

Comment (QiJune) [reply paddle/framework/tensor_util.cc]: I mean the `CopyFrom` interface is
https://github.com/PaddlePaddle/Paddle/blob/1831176764dfa288d6c4c263047c951f18b63636/paddle/framework/tensor_util.h#L33-L34

File: paddle/platform/device_context.h (hunk 5 above)
Comment (reyoung) [reply paddle/platform/device_context.h]: It is used for `DeviceCtxPool::GetByPlace()` method. This method will return the casted DeviceContext by PlaceType.

In the future, we could add a library type to this method.

File: paddle/framework/tensor_util.h [hunk 6]
```diff
@@ -207,6 +209,12 @@ inline void CopyToVector(const Tensor& src, std::vector<T>* dst) {
                src_ptr, size);
 }
 
+// Returns true if a tensor contains NAN, i.e., Not A Number.
+extern bool HasNAN(const framework::Tensor& tensor);
```
Comment (QiJune) [inline paddle/framework/tensor_util.h]: `extern` seems superfluous.

File: paddle/framework/tensor_util.cu [hunk 7]
```diff
@@ -0,0 +1 @@
+./tensor_util.cc
```
Comment (QiJune) [inline paddle/framework/tensor_util.cu]: Why not move the code in tensor_util.cc to tensor_util.h. Or we can have a tensor_util_impl.h. It's a little strange to have a symbolic link.

File: paddle/framework/tensor_util_test.cc [hunk 8]
```diff
@@ -230,6 +231,29 @@ TEST(CopyToVector, Tensor) {
...
+  Tensor src;
+  float* buf = src.mutable_data<float>({3}, CPUPlace());
+  buf[0] = 0.0;
+  buf[1] = NAN;
+  buf[2] = 0.0;
+
+  ASSERT_TRUE(HasNAN(src));
+}
+
+TEST(IsInf, CPU) {
```
Comment (QiJune) [inline paddle/framework/tensor_util_test.cc]: IsInf --> HasInf

File: paddle/framework/tensor_util.cu (hunk 7 above)
Comment (tonyyang-svail) [reply paddle/framework/tensor_util.cu]: @QiJune Reyoung probably had an offline discussion with you ready. But looks like `.cu` is necessary to make nvcc pass the compilation...

**Pull request merged**